c++/src/kj/memory.h

```diff
@@ -656,6 +698,205 @@ class SpaceFor {
+
+  inline Pin(Pin<T>&& other): t(kj::mv(other.t)) {
+  // Move T's ownership.
+  // Undefined behavior when live pointers exist, asserted when KJ_ASSERT_PTR_COUNTERS is defined.
+#ifdef KJ_ASSERT_PTR_COUNTERS
+    other.ptrCounter.assertEmpty();
+#endif
+  }
+
+  inline ~Pin() {
+  // Destroy a Pin with underlying object. 
+  // Undefined behavior when live pointers exist, asserted when KJ_ASSERT_PTR_COUNTERS is defined.
+#ifdef KJ_ASSERT_PTR_COUNTERS
+    ptrCounter.assertEmpty();
+#endif
+  }
+
+  inline T* operator->() { return get(); }
+  inline const T* operator->() const { return get(); }
+
+  inline operator Ptr<T>() { return Ptr<T>(this); }
+  // Pin<T> can be implicitly converted to Ptr<T> to obtain new pointers.
+
+  inline Ptr<T> asPtr() { return Ptr<T>(this); }
+  // Explicit convenience method to create new pointers.
+
+  template <typename U, typename = EnableIf<canConvert<T*, U*>()>>
+  inline operator Ptr<U>() { return Ptr<U>(this); }
+  // Pin<T> can be implicitly converted to pointers of compatible types.
+
+  template <typename U, typename = EnableIf<canConvert<T*, U*>()>>
+  inline Ptr<U> asPtr() { return Ptr<U>(this); }
+  // Explicit convenience method to create new pointers of compatible types.
+
+  void* operator new(size_t count) = delete;
+  void* operator new[](size_t count) = delete;
+  // Pin<T> can't be heap allocated, only local or data field usage is ok.
+
+private:
+  KJ_DISALLOW_COPY(Pin);
+
+  inline Pin(T&& t): t(kj::mv(t)) {}
+
+  T t;
+#ifdef KJ_ASSERT_PTR_COUNTERS
+  _::PtrCounter ptrCounter;
+#endif
+
+  inline T* get() { return &t; }
+  inline const T* get() const { return &t; }
+
+  template <typename>
+  friend class Ptr;
+};
+
+// =======================================================================================
+// Ptr<T>
+
+template <typename T>
+class Ptr {
+// Ptr<T> is a smart alternative to T&.
+// 
+// When used together with Pin<T> it keeps track of active pointers.
+// Asserts lifetime constraints when KJ_ASSERT_PTR_COUNTERS is defined.
+// Zero-overhead alternative for T& if KJ_ASSERT_PTR_COUNTERS is not defined.
+
+public:
+  inline ~Ptr() {
+    if (ptr == nullptr) {
+      // the value was moved out
+      return;
+    }
+#ifdef KJ_ASSERT_PTR_COUNTERS
+    counter->dec();
+#endif
+  }
+
+  Ptr(Ptr&& other) : ptr(other.ptr)
+#ifdef KJ_ASSERT_PTR_COUNTERS
+  , counter(other.counter) 
+#endif
+  {
+    other.ptr = nullptr;
+  }
+
+  Ptr(const Ptr& other) : ptr(other.ptr)
+  // Ptr<T> can be freely copied.
+#ifdef KJ_ASSERT_PTR_COUNTERS
+  , counter(other.counter) 
+#endif
+  {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+    counter->inc();
+#endif
+   }
+
+  inline void operator=(std::nullptr_t other) { 
+    if (ptr != nullptr) {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+      counter->dec();
+      counter = nullptr;
+#endif
+      ptr = nullptr;
+    }
+  }
+
+  inline T* operator->() { return get(); }
+  inline const T* operator->() const { return get(); }
```

Comment:
I think the goal, at least as I understand it, is to strongly discourage use of the bare references as opposed to completely prohibiting it. I'm happy with this, as is. If it proves to be an annoyance in various cases, then we can always add it later?

Reply:
Here's my thinking here:

- `Ptr<T>` is _the_ replacement for `T&`, obtaining it is easy (you can always clone one too).
- the more restrictive interface, the easier it is to reason about the code. Without `operator*` the value can't escape the pointer.
- I have tried converting several references upstream and so far I didn't need `operator*` (I do need Own/Ptr compatibility since we sometimes have same `T&` pointing to stack and sometimes to heap).

So yeah, I agree with @jasnell : I'd like to give it a go as is.

If it does turn out to be an annoyance, I would rather added `T& unsafeGet<T>(Ptr<T> ptr)` or something similar that is greppable ('*' isn't).

Reply:
> the more restrictive interface, the easier it is to reason about the code. Without operator* the value can't escape the pointer.

Sure they can. I can always write `*ptr.operator->()`. And if I _need_ a reference, that's exactly what I'm going to do.

I think there are two cases:

1. I don't really need a reference. In this case I, as a developer working in good faith, will try to use `kj::Ptr`. It's not the absence of `operator*` that makes me do this, it's the fact that I want to write safe code.
2. I need a reference, for some reason. I this case I will see that `*ptr` doesn't work, roll my eyes, and write `*ptr.operator->()` instead. Or worse, if I don't know that `*ptr.operator->()` works, I'll be stuck and frustrated.

I don't really buy that hiding `operator*` in itself is going to meaningfully reduce the use of bare references, but I do think it will lead to frustration.

Reply:
How about `ptr.unsafeGet()`? Could it be a reasonable compromise?

Reply:
I've been think why would one _need_ a reference, and so far interaction with 3rd party code is the only thing that comes to mind. This is definitely reasonable, but rather rare, so imo anything more explicit than `*ptr` will be a gain.

Reply:
I still don't believe that this kind of prohibition is going to be helpful but I'm willing to leave it until the first time it actually comes up.

Reply:
I've added this:

```
  inline T& asRef() { return *get(); }
  // Obtain a `T&` reference.
  // This is an unsafe operation and should be avoided unless absolutely necessary.
  // It is undefined behavior to use the reference after the object managed by this Ptr<T>
  // ceased to exist.
```